src/rest/handler.cpp

```diff
@@ -0,0 +1,473 @@
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "rest/handler.hpp"
+
+#include <string>
+
+namespace otbr {
+namespace rest {
+
+static uint16_t HostSwap16(uint16_t v)
+
+{
+    return (((v & 0x00ffU) << 8) & 0xff00) | (((v & 0xff00U) >> 8) & 0x00ff);
+}
+
+static std::string FormatBytes(const uint8_t *aBytes, uint8_t aLength)
+{
+    std::unique_ptr<char> p(new char(aLength + 1));
+
+    char *q = p.get();
+
+    for (int i = 0; i < aLength; i++)
+    {
+        snprintf(q, 3, "%02x", aBytes[i]);
+        q += 2;
+    }
+    std::string s(p.get(), aLength);
+    return s;
+}
+Handler::Handler()
+{
+    mHandlerMap.emplace("/diagnostics", &GetDiagnostic);
+    mHandlerMap.emplace("/node", &GetNodeInfo);
+    mHandlerMap.emplace("/node/state", &GetState);
+    mHandlerMap.emplace("/node/ext-address", &GetExtendedAddr);
+    mHandlerMap.emplace("/node/network-name", &GetNetworkName);
+    mHandlerMap.emplace("/node/rloc16", &GetRloc16);
+    mHandlerMap.emplace("/node/leader-data", &GetLeaderData);
+    mHandlerMap.emplace("/node/num-of-route", &GetNumOfRoute);
+    mHandlerMap.emplace("/node/ext-panid", &GetExtendedPanId);
+    mHandlerMap.emplace("/node/rloc", &GetRloc);
+}
+
+requestHandler Handler::GetHandler(std::string aPath)
+{
+    auto it = mHandlerMap.find(aPath);
+    if (it == mHandlerMap.end())
+        return ErrorHandler;
+
+    else
+    {
+        auto handler = it->second;
+        return handler;
+    }
+}
+
+void Handler::ErrorHandler(Connection &aConnection, Response &aResponse)
+{
+    aConnection.SetErrorFlag(1);
+    aConnection.SetErrorCode("no match handler");
```

Comment:
Why don't we operate on the response directly?

Reply:
Yes, It should be